File: Practice/atcoder/ABC/065/src/c.py
# 問題文読み違えてた
# 同じのが隣り合ったらダメ
# 数が同じ -> 交互に並べる
# 差が1 -> 大きいほうが外側両端をとる
# 差がそれ以上 -> はいどうしたって無理
n, m = map(int, input().split())
if n == m:
    ans = 2  # 同じ数なので入れ替えたときのやつ
    for i in range(2, n+1):
        ans *= i
        ans %= 10**9+7
        ans *= i
        ans %= 10**9+7
    print(ans)
elif abs(n - m) == 1:
    ans = 1
    for i in range(2, n+1):
        ans *= i
        ans %= 10**9+7
    for i in range(2, m+1):
        ans *= i
        ans %= 10**9+7
    print(ans)
else:
    print(0)

File: Practice/atcoder/ABC/065/src/c.rb
# 異なる数 -> 大きいほうが左右両端につく 1通り*それぞれの階乗をかけたもの(順列)
# 同じ数 -> 左右が異なるので2通り*それぞれのre
# ぜんぶ整数の掛け算なので都度割っても変わらない

n, m = gets.chomp.split.map(&:to_i)
d = 10**9+7
ans = 1
if n==m
  n.downto(1) do |i|
    # mのぶんもついでにやる
    ans*=i
    ans%=d
    ans*=i
    ans%=d
  end
  puts 2*ans%d
elsif (n-m).abs==1
  # なるべく一緒に処理する
  1.upto([n, m].max) do |i|
    if i<=n
      ans*=i
      ans%=d
    end
    if i<=m
      ans*=i
      ans%=d
    end
  end
  puts ans
else
  puts 0  
end